L1Trigger/TrackerTFP/src/TrackQuality.cc

```diff
@@ -0,0 +1,258 @@
+  }
+
+  // Helper function to convert chi2B to bin
+  int TrackQuality::toBinChi2B(double chi2B) const {
+    static const array<double, numBinsChi2B_> bins = TTTrack_TrackWord::bendChi2Bins;
+    return toBin(bins, chi2B);
+  }
+
+  // Helper function to convert chi20 to bin
+  int TrackQuality::toBinChi20(double chi20) const {
+    static const array<double, numBinsChi20_> bins = TTTrack_TrackWord::chi2RPhiBins;
+    double chi2 = chi20 * chi2rphiConv_;
+    return toBin(bins, chi2);
+  }
+
+  // Helper function to convert chi21 to bin
+  int TrackQuality::toBinChi21(double chi21) const {
+    static const array<double, numBinsChi21_> bins = TTTrack_TrackWord::chi2RZBins;
+    double chi2 = chi21 * chi2rzConv_;
+    return toBin(bins, chi2);
+  }
+
+  TrackQuality::Track::Track(const FrameTrack& frameTrack, const StreamStub& streamStub, const TrackQuality* tq) {
+    static const DataFormats* df = tq->dataFormats();
+    static const Setup* setup = df->setup();
+    frames_.reserve(setup->tqNumChannel());
+    const TrackDR track(frameTrack, df);
+    double trackChi20(0.);
+    double trackChi21(0.);
+    TTBV hitPattern(0, streamStub.size());
+    for (int layer = 0; layer < (int)streamStub.size(); layer++) {
+      const FrameStub& frameStub = streamStub[layer];
+      if (frameStub.first.isNull())
+        continue;
+      const StubKF stub(frameStub, df);
+      hitPattern.set(layer);
+      const double m20 = tq->format(VariableTQ::m20).digi(pow(stub.phi(), 2));
+      const double m21 = tq->format(VariableTQ::m21).digi(pow(stub.z(), 2));
+      const double invV0 = tq->format(VariableTQ::invV0).digi(1. / pow(stub.dPhi(), 2));
+      const double invV1 = tq->format(VariableTQ::invV1).digi(1. / pow(stub.dZ(), 2));
+      const double stubChi20 = tq->format(VariableTQ::chi20).digi(m20 * invV0);
+      const double stubChi21 = tq->format(VariableTQ::chi21).digi(m21 * invV1);
+      trackChi20 += stubChi20;
+      trackChi21 += stubChi21;
+    }
+    if (trackChi20 > tq->format(VariableTQ::chi20).range())
+      trackChi20 = tq->format(VariableTQ::chi20).range() - 1e-11;
+    if (trackChi21 > tq->format(VariableTQ::chi21).range())
+      trackChi21 = tq->format(VariableTQ::chi21).range() - 1e-11;
+    // calc bdt inputs
+    const double cot = tq->scaleCot(df->format(Variable::cot, Process::dr).integer(track.cot()));
+    const double zT = tq->scaleZT(df->format(Variable::zT, Process::kf).integer(track.zT()));
+    const int nstub = hitPattern.count();
+    const int ninterior = nstub - 1;
+    const int chi2B = tq->toBinChi2B(0);
+    const int chi20 = tq->toBinChi20(trackChi20);
+    const int chi21 = tq->toBinChi21(trackChi21);
+    // load in bdt
+    conifer::BDT<ap_fixed<10, 5>, ap_fixed<10, 5>> bdt(tq->model().fullPath());
+    // collect features and classify using bdt
+    const vector<ap_fixed<10, 5>>& output = bdt.decision_function({cot, zT, chi2B, nstub, ninterior, chi20, chi21});
+    const float mva = output[0].to_float();
```

Comment:
as long as it is not retrained to the current z0 and cot digitization we should not expect great performance. When we retrain we should change the input variables to zT(layer encoding granularity), chi2B, hitPattern, chi2rphi, chi2rz. And maybe create a version without chi2B to have something we can test in f/w.

Reply:
Sure I do think the performance won't be as good as before, but we should still expect a fake track peak about 0 and real track peak around 1 so it would be nice to double check this is the case and nothing else is way off. I agree on the retraining and will pass this info along.